contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md

```diff
@@ -0,0 +1,51 @@
+# Postpone Deletion of a Persistent Volume Claim in case It Is Used by a Pod
+
+Status: Proposal
+
+Version: GA
+
+Implementation Owner: @pospispa
+
+## Motivation
+
+User can delete a Persistent Volume Claim (PVC) that is being used by a pod. This may have negative impact on the pod.
+
+For more details see issue https://github.com/kubernetes/kubernetes/issues/45143
+
+## Proposal
+
+Postpone the PVC deletion until the PVC is not used by any pod.
+
+## User Experience
+
+### Use Cases
+
+1. User deletes a PVC that is being used by a pod. This may have negative impact on the pod. As a user, I want that any PVC deletion does not have any negative impact on any pod.
+
+## Implementation
+
+### PVC Create
+When a PVC is created finalizer is added into its metadata.
+
+### Finalizer
+Finalizer can list pods and watch for pods. It caches status of pods that use a PVC. It also watches for PVC deletes.
+
+Finalizer reacts to PVC delete in one of the below ways:
+- In case the PVC is not used by a pod: removes the finalizer information from the PVC metadata.
```

Comment:
Could you have a race condition between adding and deleting pods?

1. Pod A is currently using PVC.
2. Pod A is deleted.
3. Soon after, Pod B is created and scheduled that uses PVC.
4. PVC is deleted.
5. Informer gets Pod delete for Pod A, sees no other Pod is using the PVC in its cache, and removes finalizer.
6. Informer gets Pod add for Pod B, but finalizer is already deleted.

Reply:
PV deletion timestamp should fail setting up a PVC volume for a pod

Reply:
What if Pod B got setup with the volume before the informer receives the updates?

Reply:
I agree with @msau42  that the informer may get PVC delete before Pod B setup so in case the finalizer is removed immediately after PVC delete there is a chance for a race condition.

Currently, I can't think of other solution for the race condition than to postpone the finalizer removal from PVC for at least `deletionPostponePeriod`.

Reply:
That's not really safe either.  The controller could be down.

Isn't this best effort?  How strong of a guarantee is required?  The description above implies a strong guarantee isn't required, and is vague about what the consequences are.

Reply:
> That's not really safe either.

I agree.

> The controller could be down.

In case the controller is done the finalizer setting is not being removed from PVCs so PVCs are not being deleted and PVCs marked for deletion are piling up.

In addition, the controller is not setting finalizer for newly created PVCs.

> Isn't this best effort? How strong of a guarantee is required? The description above implies a strong guarantee isn't required, and is vague about what the consequences are.

The worst possible consequences are data lost in some of the possible scenarios that always depend on the storage type being used:
- in case the dynamic provisioning is used and reclaim policy is `Delete` the PVC deletion triggers deletion of the associated storage asset and PV.
- the same as above applies for the static provisioning and `Delete` reclaim policy.

So, IMHO, it makes sense to try to find a solution for the above race condition.